### PR TITLE
Replace component-cli with ocm

### LINKED ISSUE #457: replace "component-spec" language bindings w/ OCM-Lib
# Context / Motivation

[component-cli](https://github.com/gardener/component-cli) was deprecated (superseded by [ocm-cli](https://github.com/open-component-model/ocm) and removed from gardener-cicd (pipeline template and all pipelines) as part of [cleanup and consolidation](https://github.tools.sap/kubernetes/cicd-backlog/issues/1) activity.

`component-cli` as well as testmachinery uses golang bindings from [component-spec](https://github.com/gardener/component-spec) repository for retrieving and deserialising OCM-Component-Descriptors. Said language-bindings were also superseded by a re-implementation (in golang) in [ocm-repository](https://github.com/open-component-model/ocm).

The re-implemented ocm-lib supports handling of different versions of component-descriptors w/ a stable API (which is promised to be kept stable also for future spec-versions of OCM; there are already two spec-versions: v2 and v3 (gardener-cicd currently uses v2).

Adopting the re-implemented ocm-lib in testmachinery will allow us to archive the (currently unmaintained) "component-spec"-language-bindings and (hopefully) reduce related maintenance efforts in the future.

# Implementation Proposal

Remove dependencies from testmachinery towards https://github.com/gardener/component-spec, and port it to using [ocm-lib](https://github.com/open-component-model/ocm) instead.

### PR DESCRIPTION
**Pull request by fabianburth**
**How to categorize this PR?**

/area quality
/kind api-change

**What this PR does / why we need it**:
This PR component-cli based implementations to fetch components (to fill the location list of a testrun template) with ocm based implementations.

Thereby, it introduces breaking changes to the existing API. Primarily, so far, the component references were resolved by analysing the repository context of the referencing component and trying to fetch the referenced component from the repository specified by the last repository context. With the current ocm-spec (and the json-schema for component-descriptors), the repository context is rather informational and no longer required (see [here](https://github.com/open-component-model/ocm-spec/blob/main/doc/01-model/03-elements-sub.md#repository-contexts)).

Therefore, the repository now HAS TO be specified as a command line option (--repo) OR in an ocm config file (the location of the config file can also be specified with a command line option (--ocm-config-path). The config file also allows to specify multiple repositories (e.g. if not all referenced components are located in the same repository).

**Which issue(s) this PR fixes**:
Fixes #457 

**Special notes for your reviewer**:
Due to some dependency conflicts, I had to upgrade the docker library, which also lead to an upgrade of the go standard slices package, which had a breaking change - to be concrete, of the slices.SortFunc(). Since ocm originally imported an older version of the slices package, this lead to an issue ([here](https://github.com/open-component-model/ocm/blob/12ed2c62ffaedf033453cfec2b77b07298ccf72e/pkg/generics/set.go)  - line 64).
I resolved this issue directly in the vendored file for now.

**Release note**:

```breaking operator
replace component-cli with ocm
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,190 +1,330 @@
...
 	github.com/Masterminds/sprig v2.22.0+incompatible
 	github.com/Masterminds/sprig/v3 v3.2.3
-	github.com/argoproj/argo-workflows/v3 v3.4.13
-	github.com/bradleyfalzon/ghinstallation/v2 v2.6.0
-	github.com/fsnotify/fsnotify v1.6.0
-	github.com/gardener/component-cli v0.55.0
-	github.com/gardener/component-spec/bindings-go v0.0.98
-	github.com/gardener/gardener v1.78.0
-	github.com/ghodss/yaml v1.0.1-0.20190212211648-25d852aebe32
+	github.com/argoproj/argo-workflows/v3 v3.5.1
```
Comment (hendrikKahl) [inline go.mod:10]: Please keep the argo dependency at its original version v3.4.13. Upgrading it should be done in a separate PR incl the deployed version of the argo controller. Since it is a minor patch, it will probably include some more changes.

**Review by hendrikKahl — CHANGES_REQUESTED** [review]
Hi @fabianburth ,

thanks  a lot for the PR and the work you put in it already :) 
@dguendisch and I had a look at it and there are a few nits, where we would kindly ask you to adapt the PR:

- keep Argo WF version at its original version
- run `make format`
- fix the linter warnings in `make check`

Thanks

**Comment by hendrikKahl** [comment]
@fabianburth could you please also update the referenced Golang version to 1.21 in
- https://github.com/gardener/test-infra/tree/master/.test-defs
- https://github.com/gardener/test-infra/blob/master/Dockerfile#L16 
Thanks!

**Comment by fabianburth** [comment]
Awaiting next ocm release to be able to resolve issue stated in special notes for the reviewer

**Review by hendrikKahl — APPROVED** [review]
/lgtm

🎉
